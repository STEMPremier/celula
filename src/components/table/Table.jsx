/* Copyright 2020 Tallo Inc.,
 *
 * This file is part of Celula.
 *
 * Celula is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */
/* eslint-disable react/jsx-props-no-spreading */
import React, { useEffect } from 'react';
import PropTypes from 'prop-types';
import cx from 'classnames';

import { useTable, usePagination, useRowSelect } from 'react-table';

import Pagination from '../pagination';
import useRowSelectComponent from './useRowSelectComponet';
import { functionOrUndef } from '../../utils/propValidators';

import './table.less';

/**
 * `Tables` display information in a grid-like format of rows and columns. They organize information in a way thatâ€™s easy to scan, so that users can look for patterns and insights. Tables can contain interactive components (such as chips, buttons, or menus), non-interactive elements (such as badges).
 */
/* ### *Mobile Tables*
 * These list can support up to one inline link with additional actions placed within an ellipse menu in stacked tables. The purpose of this is to retain a relatively short row height. These lists are also stacked directly next to one another vertically allowing for no spacing. Text truncation is an option if desired.
 */
const Table = ({
  className,
  clickable,
  clickFn,
  columns,
  data,
  fetchData,
  pageCount: serversidePageCount, // for serverside pagination
  pageSize: clientsidePageSize, // for clientside pagination
  selectable,
  selectionHeaderFn,
  selectionFn,
  style,
}) => {
  const options = {
    columns,
    data,
  };

  let controlled = false;

  // If we have both pageCount and fetchData, we assume server-side pagination.
  if (serversidePageCount && fetchData) {
    options.manualPagination = true;
    options.pageCount = serversidePageCount;

    controlled = true;
  }

  const {
    getTableProps,
    getTableBodyProps,
    headerGroups,
    prepareRow,
    page,
    canPreviousPage,
    canNextPage,
    pageCount,
    gotoPage,
    nextPage,
    previousPage,
    setPageSize,
    state: { pageIndex, pageSize },
  } = useTable(
    options,
    usePagination,
    useRowSelect,
    useRowSelectComponent(selectionHeaderFn, selectionFn),
  );

  // When (if) the prop pageSize changes, tell the Table
  useEffect(() => setPageSize(clientsidePageSize), [clientsidePageSize]);

  // When (if) any one of (prop) fetchData, (the react-table states) pageIndex, or pageSize changes, fetch new data.
  useEffect(() => {
    if (controlled) fetchData({ pageIndex, pageSize });
  }, [fetchData, pageIndex, pageSize]);

  const classes = cx(
    'ce-table',
    {
      'ce-table__clickable': clickable,
      'ce-table__selectable': selectable,
    },
    className,
  );

  const navToPage = pg => gotoPage(pg - 1);

  return (
    <>
      <div className={classes} style={style} {...getTableProps()}>
        {/* table header */}
        <div className="ce-table__header" role="rowgroup">
          {headerGroups.map(headerGroup => (
            <div
              key={headerGroup.id}
              className="ce-table__header--group"
              {...headerGroup.getHeaderGroupProps()}
            >
              {headerGroup.headers.map(column => (
                <div
                  key={column.id}
                  className="ce-table__heading"
                  {...column.getHeaderProps()}
                >
                  {column.render('Header')}
                </div>
              ))}
            </div>
          ))}
        </div>

        {/* table body */}
        <div className="ce-table__body" {...getTableBodyProps()}>
          {page.map(row => {
            prepareRow(row);

            const rowProps = {
              className: `ce-table__row${
                row.isSelected ? ' ce-table__row--selected' : ''
              }`,
              ...row.getRowProps(),
              onClick: clickable
                ? event => clickFn(row.original, event)
                : () => {},
              tabIndex: clickable ? '0' : null,
            };

            return (
              <div key={row.id} {...rowProps}>
                {row.cells.map(cell => (
                  <div
                    key={cell.id}
                    className="ce-table__cell"
                    {...cell.getCellProps()}
                  >
                    <div className="ce-table__cell-heading">
                      {typeof cell.column.Header === 'string'
                        ? `${cell.render('Header')}:`
                        : cell.render('Header')}
                    </div>
                    <div className="ce-table__cell-body">
                      {cell.render('Cell')}
                    </div>
                  </div>
                ))}
              </div>
            );
          })}
        </div>
      </div>
      {pageCount > 1 && (
        <Pagination
          canPreviousPage={canPreviousPage}
          canNextPage={canNextPage}
          currentPage={pageIndex}
          gotoPage={navToPage}
          nextPage={nextPage}
          pageCount={pageCount}
          prevPage={previousPage}
        />
      )}
    </>
  );
};

Table.propTypes = {
  /**
   * A class name, or string of class names, to add to the `Table`.
   */
  className: PropTypes.string,
  /**
   * An array for the column definition of your `Table`.
   */
  columns: PropTypes.arrayOf(PropTypes.object).isRequired,
  /**
   * An array of data (objects) for the `Table`.
   */
  data: PropTypes.arrayOf(PropTypes.object).isRequired,
  /**
   * A function the `Table` uses to fetch data. This is only used when utilizing server-side pagination.
   */
  fetchData: functionOrUndef,
  /**
   * Used to tell the `Table` how many pages are there when utilizing server-side pagination.
   */
  pageCount: PropTypes.number,
  /**
   * The number of items per page in the `Table`.
   */
  pageSize: PropTypes.number,
  /**
   * Make the rows in the `Table` clickable.
   */
  clickable: PropTypes.bool,
  /**
   * A function to trigger when clicking on a row in the `Table`.
   * The function accepts 2 arguments, `row` and `event`, in that order: `(row, event) => {}`.
   *
   * @param {row} object - The row object for this row of the `Table`.
   * @param {event} object - The event triggered by the click on the row.
   */
  clickFn: PropTypes.func,
  /**
   * Adds a checkbox in the `Table`, as the first column.
   */
  selectable: PropTypes.bool,
  /**
   * A function to trigger when clicking on the checkbox in the header row in the `Table`.
   * The function accepts 2 arguments, `row` and `event`, in that order: `(row, event) => {}`.
   *
   * @param {row} object - The row object for this row of the `Table`.
   * @param {event} object - The event triggered by the click on the row.
   */
  selectionHeaderFn: PropTypes.func,
  /**
   * A function to trigger when clicking on the checkbox in a row in the `Table`.
   * The function accepts 2 arguments, `rows` and `event`, in that order: `(row, event) => {}`.
   *
   * @param {rows} object - The row object for this row of the `Table`.
   * @param {event} object - The event triggered by the click on the row.
   */
  selectionFn: PropTypes.func,
  /**
   * Any inline styles you would like to add to the `Table`. See the React [docs](https://reactjs.org/docs/faq-styling.html) for more.
   */
  style: PropTypes.object, // eslint-disable-line react/forbid-prop-types
};

Table.defaultProps = {
  className: '',
  fetchData: undefined,
  pageCount: 0,
  pageSize: 10,
  clickable: false,
  clickFn: () => {},
  selectable: false,
  selectionHeaderFn: () => {},
  selectionFn: () => {},
  style: {},
};

export default Table;
/* eslint-enable react/jsx-props-no-spreading */
