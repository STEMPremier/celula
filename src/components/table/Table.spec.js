/* eslint-disable react/jsx-filename-extension */
import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';

import Table from '.';

describe('<Table />', () => {
  describe('render()', () => {
    test('renders the component', () => {
      const wrapper = shallow(
        <Table
          label="This is a label"
          name="Name"
          headings={['Heading 1', 'heading 2']}
          rows={[
            { data: ['row 1 cell 1', 'row 1 cell 2'] },
            { data: ['row 2 cell 1', 'row 2 cell 2'] },
          ]}
        />,
      );
      expect(toJson(wrapper)).toMatchSnapshot();
    });
  });
});
/* eslint-enable react/jsx-filename-extension */
