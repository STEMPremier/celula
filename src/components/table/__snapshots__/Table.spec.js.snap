// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Table /> render() renders the component 1`] = `
<div
  className="ce-table"
  headings={
    Array [
      "Heading 1",
      "heading 2",
    ]
  }
  label="This is a label"
  name="Name"
  onChange={false}
  rows={
    Array [
      Object {
        "data": Array [
          "row 1 cell 1",
          "row 2 cell 2",
        ],
      },
      Object {
        "data": Array [
          "row 2 cell 1",
          "row 2 cell 2",
        ],
      },
    ]
  }
  selectable={false}
>
  <div
    className="ce-table__container-desktop"
  >
    <div
      className="ce-table__header-row"
    >
      <div
        className="ce-table__header-cell"
      >
        Heading 1
      </div>
      <div
        className="ce-table__header-cell"
      >
        heading 2
      </div>
    </div>
    <div
      className="ce-table__data-row null"
    >
      <div
        className="ce-table__data-cell "
      >
        row 1 cell 1
      </div>
      <div
        className="ce-table__data-cell "
      >
        row 2 cell 2
      </div>
    </div>
    <div
      className="ce-table__data-row null"
    >
      <div
        className="ce-table__data-cell "
      >
        row 2 cell 1
      </div>
      <div
        className="ce-table__data-cell "
      >
        row 2 cell 2
      </div>
    </div>
  </div>
  <div
    className="ce-table__container-mobile"
  >
    <div
      className="ce-table__mobile-row"
    >
      <div
        className="ce-table__mobile-card-container null"
      >
        <div
          className="ce-table__mobile-data-container"
        >
          <div
            className="ce-table__mobile-headings"
          >
            <span>
              <strong>
                Heading 1
                :
              </strong>
            </span>
            <span>
              <strong>
                heading 2
                :
              </strong>
            </span>
          </div>
          <div
            className="ce-table__mobile-card-data"
          >
            <div
              className="ce-table__mobile-card-row-data"
            >
              row 1 cell 1
            </div>
            <div
              className="ce-table__mobile-card-row-data"
            >
              row 2 cell 2
            </div>
          </div>
        </div>
      </div>
      <div
        className="ce-table__mobile-card-container null"
      >
        <div
          className="ce-table__mobile-data-container"
        >
          <div
            className="ce-table__mobile-headings"
          >
            <span>
              <strong>
                Heading 1
                :
              </strong>
            </span>
            <span>
              <strong>
                heading 2
                :
              </strong>
            </span>
          </div>
          <div
            className="ce-table__mobile-card-data"
          >
            <div
              className="ce-table__mobile-card-row-data"
            >
              row 2 cell 1
            </div>
            <div
              className="ce-table__mobile-card-row-data"
            >
              row 2 cell 2
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
